### PR TITLE
Implement SAVE-INPUT and RESTORE-INPUT

### PR DESCRIPTION
**Pull request by ellerh**
There used to be primitive tokens ID_SAVE_INPUT and ID_RESTORE_INPUT
but those weren't used.  Saving/restoring positions in files is
somewhat involved so I decided to it in Forth.  To support this, I
re-purposed the codes of ID_SAVE_INPUT and ID_RESTORE_INPUT to
save/store the current line number (ID_SOURCE_LINE_NUMBER_FETCH, and
ID_SOURCE_LINE_NUMBER_STORE).  Those can also be used for something
like C's __LINE__ macro.

* fth/save-input.fth: New file.
* fth/loadp4th.fth: Load it.
* fth/system.fth (D=): New. Needed to compare file positions.
* fth/t_corex.fth: Add simple tests.

* csrc/pf_guts.h (ID_SOURCE_LINE_NUMBER_FETCH,
ID_SOURCE_LINE_NUMBER_STORE): Renamed from ID_SAVE_INPUT and
ID_RESTORE_INPUT.

* csrc/pf_inner.c (ID_SOURCE_LINE_NUMBER_FETCH,
ID_SOURCE_LINE_NUMBER_STORE): Implement.
(ID_SAVE_INPUT): Deleted. It's now in Forth.

* csrc/pfcompil.c (pfBuildDictionary): Define SOURCE-LINE-NUMBER@ and
 SOURCE-LINE-NUMBER!.

### CODE REVIEW

File: fth/save-input.fth [hunk 1]
```diff
@@ -0,0 +1,67 @@
+\ SAVE-INPUT and RESTORE-INPUT
+
+anew task-save-input.fth
+
+private{
+
+: save-buffer ( -- column source-id 2 ) >in @ source-id 2 ;
+
+: restore-column ( column -- flag )
+  source nip over < if drop true exit then
```
Comment (philburk) [inline fth/save-input.fth]: Could we use an ELSE instead of EXIT THEN?

File: fth/save-input.fth [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+\ Source      Stack
+\ EVALUATE    >IN  SourceID=(-1)  2
+\ keyboard    >IN  SourceID=(0)   2
+\ file        >IN  lineNumber filePos  SourceID=(fileID) 5
+: SAVE-INPUT ( -- column {line filepos}? source-id n )
+  source-id case
+    -1 of save-buffer endof
+    0  of save-buffer endof
+    drop save-file exit
```
Comment (philburk) [inline fth/save-input.fth]: Do we need this exit? The exits make it a little hard to follow the flow.

File: fth/save-input.fth [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  source-line-number@
+  line-start-position
+  source-id
+  5
+;
+
+: restore-file ( column line filepos:ud -- flag )
+  source-id reposition-file if 2drop true exit then
+  source-line-number!
+  refill 0= if drop true exit then
```
Comment (philburk) [inline fth/save-input.fth]: I have been very inconsistent with style. I would like to move towards capitalizing IF ELSE and THEN and lining them up. Also using 4 space indents. Maybe you could help me with a style guide.

    foo @ >
    IF  drop foo @
    ELSE dup swap drop
    THEN

**Comment by philburk** [comment]
Fantastic. I really appreciate you adding all these great new features. Very well implemented. I just have a few minor comments.

File: fth/system.fth [hunk 4]
```diff
@@ -360,6 +360,10 @@
         2* swap
 ;
 
+: D= ( xd1 xd2 -- flag )
+	rot = -rot = and
```
Comment (ellerh) [inline fth/system.fth]: I suppose shorter is faster, so yes.  Hmm, but it would have to be `D- D0=` and we don't have `D0=`. I'm also bit usure what happens on underflow.

File: fth/save-input.fth (hunk 3 above)
Comment (ellerh) [inline fth/save-input.fth]: In commit https://github.com/philburk/pforth/pull/25/commits/09d08b416a8385ad4e8395044d717809e18bafa7, I re-indented with 4 spaces. 

For my own projects I like to imitate the style that Mitch Bradley uses in [Open Firmware](https://github.com/MitchBradley/openfirmware), except that I use 2 spaces for indentation instead of 3. Mitch Bradley's code is IMO very clear.  The style in pForth is of course different and I'm happy to use whatever you prefer.

I'm now just writing down some of my thoughts on style (feel free to disagree :-):
* I'm no fan of uppercase keywords, as it's harder to type and my editor already highlights the keywords. I write almost everything in lowercase. An exception are constants like `SEEK_CUR` that are the same as in C. Also in comments, an uppercase name refers to a variable (to avoid quoting).
* I use `-` to combine words like `foo-bar` instead of `foo.bar` or `foo_bar`.  As all the standard words use `-`  this looks more consistent. 
* As you've noticed, I use EXIT frequently. IMO that's more readable than nested IFs.  I even prefer EXIT to terminate loops over WHILE.  Loop constructs like `BEGIN  ... WHILE  ... WHILE ... REPEAT THEN` are confusing. (And difficult to recongnize for editors.)
* Usually I avoid non-letter characters like `>$&#%` in names and use words instead. Exceptions
are `foo?` which is a predicate, `foo>bar` which means "move or convert `foo` to `bar`" and `/foo` is the size (usually in bytes) of struct `foo`.  In the past I used `(foo)` for internal/private helpers, but these days I prefer `%foo` as it is shorter and `%%foo` looks better than `((foo))`.
* I'm still undecided whether `;` should be on same line as the last word in a colon definition or on it's own line.

File: fth/save-input.fth (hunk 2 above)
Comment (ellerh) [inline fth/save-input.fth]: Remeber that ENDCASE drops one item from the stack. So one could put a dummy value there first
with `drop save-file 0` or `>r save-file r>` but IMO neither is better than EXIT.

File: fth/system.fth (hunk 4 above)
Comment (philburk) [inline fth/system.fth]: Right. D0= missing. Never mind.

File: fth/save-input.fth (hunk 3 above)
Comment (philburk) [inline fth/save-input.fth]: Thanks for the input. I want pForth to be more consistent with my other Forth projects like HMSL and JForth.
https://github.com/philburk/hmsl/blob/master/hmsl/fth/midi_parser.fth

I'd love to continue this thread elsewhere. Are you on the pforthdev list?
https://groups.google.com/forum/#!forum/pforthdev

Comment (ellerh) [inline fth/save-input.fth]: > Are you on the pforthdev list?

Now  I am.
